lib/useful.ex

```diff
@@ -106,4 +116,73 @@ defmodule Useful do
   defp key_to_atom({key, value}) do
     {String.to_atom("#{key}"), value}
   end
+
+  @doc """
+  `typeof/1` returns the type of a vairable.
+  Inspired by stackoverflow.com/questions/28377135/check-typeof-variable-in-elixir
+
+  ## Examples
+
+      iex> Useful.typeof(:atom)
+      "atom"
+
+      iex> bin = "hello"
+      iex> Useful.typeof(bin)
+      "binary"
+
+      iex> bitstr = <<1::3>>
+      iex> Useful.typeof(bitstr)
+      "bitstring"
+
+      iex> Useful.typeof(:true)
+      "boolean"
+
+      iex> pi = 3.14159
+      iex> Useful.typeof(pi)
+      "float"
+
+      iex> fun = fn (a, b) -> a + b end
+      iex> Useful.typeof(fun)
+      "function"
+
+      iex> Useful.typeof(&Useful.typeof/1)
+      "function"
+
+      iex> int = 42
+      iex> Useful.typeof(int)
+      "integer"
+
+      iex> list = [1,2,3,4]
+      iex> Useful.typeof(list)
+      "list"
+
+      iex> map = %{:foo => "bar", "hello" => :world}
+      iex> Useful.typeof(map)
+      "map"
+
+      iex> Useful.typeof(nil)
+      "nil"
+
+      iex> pid = spawn(fn -> 1 + 2 end)
+      iex> Useful.typeof(pid)
+      "pid"
+
+      iex> port = Port.open({:spawn, "cat"}, [:binary])
+      iex> Useful.typeof(port)
+      "port"
+
+      iex> ref = :erlang.make_ref
+      iex> Useful.typeof(ref)
+      "reference"
+
+      iex> tuple = {:name, "alex"}
+      iex> Useful.typeof(tuple)
+      "tuple"
+  """
+  types =
+    ~w[boolean binary bitstring float function integer list map nil pid port reference tuple atom]
+
+  for type <- types do
+    def typeof(x) when unquote(:"is_#{type}")(x), do: unquote(type)
```

Comment:
I was wondering if the function would break if used with a [custom type](https://elixir-lang.org/getting-started/typespecs-and-behaviours.html#defining-custom-types):

```elixir
@type year :: integer
```
but `typeof` will return the type itself instead of the alias (ie `integer` if used with `year`). I can't see a way to returns the alias type but I don't think this is an issue.
Otherwie code, tests and docs look good to me :+1:

Reply:
@SimonLab _great_ question! 💡 
perhaps we need a fallback `unknown` response for when the `type` is not in the list on `L183` 💭

Reply:
I think the list can be extended when required, I think there are a lot more defined types in Elixir: https://hexdocs.pm/elixir/1.12/typespecs.html.
We could definitely create a fallback response :+1:

Reply:
Good catch. Thanks for adding the catch-all `"unknown"` 👍

Reply:
Just spotted that we don't have test coverage of this ... 🤷‍♂️ 
I cannot think of how to test it it. Can you? 💭